File: offer_1_0054/java/offer_1_0054.java
import java.util.*;
public class offer_1_0054 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public int kthLargest(TreeNode root, int k) {
            //求最大的k个节点 采用先右根左的顺序 保证从大到小
            target = 0;
            TreeNode targetNode = findKthLarge(root,k);
            if(targetNode == null)
                return Integer.MIN_VALUE;
            return targetNode.val;
        }
        private int target;
        private TreeNode findKthLarge(TreeNode root,int k){
            if(root != null){
                TreeNode rNode = findKthLarge(root.right,k);
                if(rNode != null)
                    return rNode;
                target++;
                if(k == target)
                    return root;
                TreeNode lNode = findKthLarge(root.left,k);
                if(lNode != null)
                    return lNode;
            }
            return null;
        }
        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        offer_1_0054.Solution solution = new offer_1_0054().new Solution();
        offer_1_0054.TreeNode root = new offer_1_0054().new TreeNode(3);
        offer_1_0054.TreeNode left = new offer_1_0054().new TreeNode(1);
        offer_1_0054.TreeNode right = new offer_1_0054().new TreeNode(4);
//        offer_1_0054.TreeNode ll = null;
        offer_1_0054.TreeNode lr = new offer_1_0054().new TreeNode(2);
//        offer_1_0054.TreeNode rl = new offer_1_0054().new TreeNode(3);
//        offer_1_0054.TreeNode rr = new offer_1_0054().new TreeNode(6);
//        left.left = ll;
        left.right = lr;
//        right.left = rl;
//        right.right = rr;
        root.left = left;
        root.right = right;
//        root = [2,1,3], p = 1
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));
        //root = [3,1,4,null,2], k = 1
        int k = 1;
        int res = solution.kthLargest(root,k);
        System.out.println(res);
        //result
        //4
    }
}

File: offer_1_0054/cpp/offer_1_0054.cpp
//
// Created by wzy on 2022/4/6.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    int kthLargest(TreeNode* root, int k) {
        //求最大的k个节点 采用先右根左的顺序 保证从大到小
        int target = 0;
        TreeNode* targetNode = findKthLarge(root,k,target);
        if(!targetNode)
            return INT_MIN;
        return targetNode->val;
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
private:
    TreeNode* findKthLarge(TreeNode* root,int k,int& target){
        if(root){
            TreeNode* rNode = findKthLarge(root->right,k,target);
            if(rNode)
                return rNode;
            target++;
            if(k == target)
                return root;
            TreeNode* lNode = findKthLarge(root->left,k,target);
            if(lNode)
                return lNode;
        }
        return nullptr;
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
    TreeNode *root = new TreeNode(3);
    TreeNode *left = new TreeNode(1);
    TreeNode *right = new TreeNode(4);
//    TreeNode* ll = nullptr;
    TreeNode* lr = new TreeNode(2);
//    TreeNode* rl = new TreeNode(3);
//    TreeNode* rr = new TreeNode(6);
    root->left = left;
    root->right = right;
//    left->left = ll;
    left->right = lr;
//    right->left = rl;
//    right->right = rr;
    vector<string> rootArr = Solution().levelOrderStr(root);
    printVec(rootArr);
    //root = [3,1,4,null,2], k = 1
    int k = 1;
    int res = Solution().kthLargest(root,k);
    cout<<res<<endl;
    //result
    //4
}